### DISCUSSION TITLE
Getting rid of duplicated journalctl logs

### DISCUSSION THREAD

**Discussion by 4JX** [aksiksi/compose2nix#72] [Q&A] [answered]
I'm guessing this is a byproduct of how nixos itself handles containers by assigning them to systemd units, but ever since I switched backends* from podman to docker I've been having this annoying issue where practically every log line is printed both by the container and the service managing it:

![image](https://github.com/user-attachments/assets/9a73ba57-7c50-4585-a81a-a2e9ae7d4ccc)

*I don't recall it being the case with podman, but I'm not 100% sure.

Is there a way to disable one of the two log streams? Preferably the systemd unit one, as that doesn't seem to preserve color.

**Comment by aksiksi** [comment]
`compose2nix` sets the [`log-driver`](https://search.nixos.org/options?channel=24.11&from=0&size=50&sort=relevance&type=packages&query=log-driver) to `journald` on each OCI container. This results in a `--log-driver` flag being passed to the `docker run ...` command executed by the OCI container systemd service. One of these logs is definitely due to this flag.

For Docker specifically, you can also configure a log driver at the daemon/engine level ([ref](https://docs.docker.com/engine/logging/configure/#configure-the-default-logging-driver)). This is controlled by the [`virtualisation.docker.logDriver`](https://search.nixos.org/options?channel=24.11&show=virtualisation.docker.logDriver&from=0&size=50&sort=relevance&type=packages&query=logDriver) setting, which defaults to `journald`. Could you try setting this log driver to `none`?

**Reply by 4JX** [reply]
Apologies for the delay in answering. 

I'd already tried `virtualisation.docker.logDriver = "none"` but that had no effect. Setting the service specific config to `json-file` or `none` eliminates the "container" logs and keeps the `*-start` service ones.

**Reply by aksiksi** [reply, marked as answer]
I *think* I see what the problem is.

For Podman, the container runs in detached mode, which means container output does not go to stdout: https://github.com/NixOS/nixpkgs/blob/b47fd6fa00c6afca88b8ee46cfdb00e104f50bca/nixos/modules/virtualisation/oci-containers.nix#L335

I guess this can only be done with Podman because it is natively integrated with systemd. That is, systemd events will propagate to Podman and vice-versa. 

I would recommend filing an upstream issue to ask about this, and whether it would be possible to add an option to suppress stdout in favor of the `log-driver` setting. One idea could be to disable stdout logging for Docker if `log-driver == "journald"`.

For now, your only option is to disable logging at the service level.
